services/QuillLMS/app/lib/quill_authentication.rb

```diff
@@ -68,10 +70,22 @@ def sign_in(user)
     @current_user = user
   end
 
+  def demo_id=(demo_id)
+    session[:demo_id] = demo_id
+    if demo_id
+      Analyzer.new.track(current_user, SegmentIo::BackgroundEvents::VIEWED_DEMO)
+      @current_user = User.find(session[:demo_id])
+    else
+      @current_user = User.find(session[:user_id])
+    end
+  end
+
   def preview_student_id=(student_id)
     session[:preview_student_id] = student_id
     if student_id
       @current_user = User.find(session[:preview_student_id])
+    elsif session[:demo_id].present?
+      @current_user = User.find(session[:demo_id])
```

Comment:
This is here because if the user is previewing as a student from inside the Demo account, we need to bring them back into the demo account when they exit the student preview.

We can longer bring them back to the demo account through `session[:user_id]` because that field now contains the ID of the actual teacher, not the demo teacher.

You can think of it as a double-nested action. A teacher jumps into the Demo account, then from the Demo account jumps into a student account. When they jump out of the student account, we want to let them jump out one level back into the Demo account, not two levels back into their own account.

Reply:
That makes sense! Thanks for clarifying.